File: 贪心算法/middle/45.跳跃游戏-ii.cpp
#include <iostream>
#include <vector>
#include <string>
#include <stack>
#include <queue>
#include <algorithm>
#include <unordered_set>
#include <unordered_map>
#include <math.h>

using namespace std;
/*
 * @lc app=leetcode.cn id=45 lang=cpp
 *
 * [45] 跳跃游戏 II
 */

// @lc code=start
class Solution
{
public:
    int jump(vector<int> &nums)
    {
        int current = 0;
        int next = 0;
        int count = 0;

        // nums.size() - 1,标识走到倒数第二个位置，再走一步一定能走到
        for (int i = 0; i < nums.size() - 1; i += 1)
        {
            next = max(next, nums[i] + i);

            // 下标走到当前的最大范围，则需要步数+1
            // 走到最后nums.size() - 1的位置，如果i != current，表明当前这一步已经走到了，不需要+1
            if (i == current)
            {
                count++;
                // 继续扩大当前的最大范围
                current = next;
            }
        }

        return count;
    }
};
// @lc code=end

File: 贪心算法/middle/45.跳跃游戏-ii.ts
/*
 * @lc app=leetcode.cn id=45 lang=typescript
 *
 * [45] 跳跃游戏 II
 */

// @lc code=start
function jump(nums: number[]): number {
  let count = 0
  // 当前范围起始下标
  let startIndex = 0
  // 当前范围结尾下表
  let endIndex = 0

  if (nums.length === 1) {
    return 0
  }

  // 结尾下标如果超过数组最后一个元素，则到达结尾，直接返回步数
  while (endIndex < nums.length - 1) {
    let i = startIndex
    let j = endIndex
    // 遍历当前范围内每个元素能到达的下一个范围
    for (; i <= j; i++) {
      let coverIndex = nums[i] + i

      if (coverIndex > endIndex) {
        // 纪录最大范围，下一次遍历使用
        endIndex = coverIndex
      }
    }
    // 当前遍历的结束时j，name下一次便利从j+1到新的endIndex
    startIndex = j + 1
    // 记录一次步数
    count++
  }

  return count
};
// @lc code=end

